### ISSUE TITLE
[popup] Bikeshed `popover=hint`

### ISSUE THREAD

**Issue by mfreed7** [openui/open-ui#532]
It has been mentioned often that `popup=hint` and `popup=async` aren't great at describing what they do. Perhaps we should pick better names?

For `popup=hint`, this is almost always referred to as a "tooltip". Should we use `popup=tooltip`? That does seem to "bake in" one particular use case, and there might be others. Then again, because of the strong prior knowledge about how a "tooltip" behaves, it also makes it easy to quickly understand the behavior. Is there another word that describes this behavior but is more generic?

For `popup=async`, this is variously used for things that are called "toasts", "notifications", "real-time notifications", "alerts", "snackbars", etc. The behavior is something that asynchronously (hence the name) shows up, does not change focus, does not close other elements or light dismiss, and sometimes times out and disappears. What name better explains this? Or maybe `async` does the job.

**Comment by mfreed7** [comment]
Note this comment by @domenic https://github.com/openui/open-ui/issues/495#issue-1171650620 which suggests `popup=manual` instead of `async`. Makes sense to me.

I also thought about `popup=sticky` since the behavior is "sticky" - it doesn't go away when other things open or when you click/move to other elements. Unfortunately, that's likely too close to `position:sticky` which is different. But maybe something similar?

**Comment by scottaohara** [comment]
i still favor `popup=hint`, as i'd submit 'hint' is the more generic term that things like 'tooltip' or browser validation error messages would fall under.  but if those are the only use cases we see for this type of popup, then i guess it wouldn't be the worst thing to have this particular value be so very specific while the others are more generic names?

**Comment by mfreed7** [comment]
Let's use this issue to discuss specifically `popup=hint`. From [this issue](https://github.com/openui/open-ui/issues/495), here are a few suggestions:

- `popup=hint`
- `popup=weak`
- `popup=light`
- `popup=transitory`
- `popup=transient`
- `popup=ancillary`

Other ideas?

**Comment by mfreed7** [comment]
Please either vote here using emoji (see below), or suggest something better. Thanks!

* `popup=hint` 👍 
* `popup=weak` 😄
* `popup=light` 🎉  
* `popup=transitory` ❤️ 
* `popup=transient` 🚀 
* `popup=ancillary` 👀

**Comment by VicGUTT** [comment]
I'm on par with @scottaohara's comment above.

A quick Duckduckgo'ing of a definition for "tooltip" and a quick read of [Wikipedia's Tooltip page](https://en.wikipedia.org/wiki/Tooltip) yields words like: hint, info, description and help.

So depending on what we want the value to convey, here's a small list of alternatives I've put together:

**Meaning**
- hint
- tip
- title _(akin to the existing "title" attribute)_
- description
- help
- info

**Behavior**
- light
- temporary _(instead of "transitory" as I think it's a more common and easily understandable word)_

**Action**
- pointer _(hovered, tapped, long touched, stylus'd)_

**No! God! No! God! Please! No! No! No! Nooooooooooooooooooo!** 😅 (-> https://media.giphy.com/media/d10dMmzqCYqQ0/giphy-downsized-large.gif)
- transient
- ancillary

-> Explanation: I'd find it unfortunate if I'd have to open a dictionary every time I see those words _(specially as a none native English speaker)_.

---

Based on the list above, my picks in order would be: hint/light, tip, title, then help/info/description

**Comment by VicGUTT** [comment]
Also thought of "delay" but, nah...

**Comment by hidde** [comment]
Like @VicGUTT , I find ‘temporary’ easier to understand than ‘transitory’ and would also need to look up and therefore be against ‘transient’ and ‘ancillary’ (all as non native speaker).

I voted for ‘hint’, because it personally gives me associations with the kind of thing popup=hint would do. I like ‘tooltip’ too, but can understand people would find it too specific.

**Comment by VicGUTT** [comment]
Same with regards to "tooltips". I think it's fine but worried it could get some backslashes similar to how the "masonry" keyword for grid did.

**Comment by bkardell** [comment]
Just setting up a polling mechanism since the options resolved don't match those of the original emoji poll

- hint 👍 
- title 😄 
- light 💡 
- info ℹ️ 
- advisory ❗

**Comment by mfreed7** [comment]
Thank you @bkardell for setting up the poll.

Folks, please vote! And encourage other people to vote. We need to resolve this issue soon and pick a name. I'd ideally like to resolve on a name at the August 4 meeting, if possible.

**Comment by Westbrook** [comment]
`hint` is great when open/close semantics ONLY are a managed. 

If `title` was used, I’d expect there to be more accessible semantics exposed by default. 

`light` implies `dark` which seems incorrect for “popups”. 

`info` is close to `hint` and could be a close second. 

`advisory` is just long enough to be wrong with no further explanation.

**Comment by domenic** [comment]
None of these seem to be about behavior (which, IIUC, is force-closing others of its type, and also responding to various dismissals). All of them seem to be about semantics. It feels like the feedback on *why* to rename wasn't really taken into account, and the group just came up with a bunch of new semantic-name synonyms?

**Comment by scottaohara** [comment]
what about "quick" or "eager" (to reuse a value from the `loading` attribute)?  

something to indicate that the `popup` will display by a user focusing or hovering the associated element (`auto` is out because we're already using it - and it's also not really automatic because there is still something that must be done to make it show).

**Comment by WickyNilliams** [comment]
I personally think `tooltip` is best. I feel like there's a universal understanding of what this means, and what behaviour is associated with it. Nobody seems to have come up with other use cases for this behaviour, so why not the obvious choice?

**Comment by mfreed7** [comment]
> None of these seem to be about behavior (which, IIUC, is force-closing others of its type, and also responding to various dismissals). All of them seem to be about semantics. It feels like the feedback on _why_ to rename wasn't really taken into account, and the group just came up with a bunch of new semantic-name synonyms?

No, we did discuss this in the meeting for sure. The `light` suggestion was discussed at some length, with "light" as in "lightweight". The problem is that the [behavior](https://github.com/openui/open-ui/blob/main/research/src/pages/popup/popup.research.explainer.mdx#classes-of-top-layer-ui) is more than can be described in a single word. The behavior, though, is **well described** by the canonical example of that behavior, the "tooltip". The word "hint" is close, and is less directly tied to a particular UI pattern, so maybe ok. Imagine if I asked you to pick a single word to describe the *behavior* of a [computer](https://en.wikipedia.org/wiki/Computer). It can be done, but the most descriptive word is "computer", since that conveys what a computer does better than any single word or short phrase, even if it's only one example embodiment of that behavior.

I'm really not arguing against your point, I just think that no great alternatives have emerged. So far for behavioral words we have: `light`, `temporary`, `transitory`, `transient`, `ancillary`, `quick`, and `eager`.

The last two were recently added, and while I admit they're better, they also don't describe the one-at-a-time and light dismiss behaviors of this type of pop-up. The rest (which we discussed) were fairly universally hated. None of them really describes the behavior, and all (I can almost guarantee) will end up causing developer confusion. Whereas people know what a "tooltip" or maybe a "hint" is, and they have (correct) expectations about how it behaves.

More/better suggestions are appreciated! The resolution was attempting to force this discussion, and it seems to be working.

**Comment by annevk** [comment]
What I miss here is consideration for `::tooltip`: https://github.com/w3c/csswg-drafts/issues/8930. I feel pretty strongly these features need to be named consistently. Perhaps the CSS WG needs to rename to `::hint`. I'll leave that to you all to make the case to them, but we shouldn't end up with two separate names just because a different group of people worked on it.

**Comment by mfreed7** [comment]
> What I miss here is consideration for `::tooltip`: [w3c/csswg-drafts#8930](https://github.com/w3c/csswg-drafts/issues/8930). I feel pretty strongly these features need to be named consistently. Perhaps the CSS WG needs to rename to `::hint`. I'll leave that to you all to make the case to them, but we shouldn't end up with two separate names just because a different group of people worked on it.

I tend to agree with this viewpoint. As the person who wrote the [popover=hint explainer](https://open-ui.org/components/popover-hint.research.explainer/), I really struggled to describe any use case besides a "tooltip". It actually made the explainer harder to write, because I had to say "here's the use case, tooltip, and from now on in the doc I'll call it a 'tooltip', but the feature is called 'hint'". @annevk's point is a good one that there's already a concept on the platform called explicitly [`::tooltip`](https://github.com/w3c/csswg-drafts/issues/8930). Feels like maybe we should rename to `popover=tooltip`. @domenic you had opinions last time we discussed this - WDYT?

**Comment by domenic** [comment]
> @domenic you had opinions last time we discussed this - WDYT?

I think we're running into a conflict between two things:

1. Consistency between HTML and CSS
2. The fact that HTML is, in its platonic ideal, about behavior and semantics, whereas CSS, in its platonic ideal, is about styling.

So CSS folks wanting a name like "tooltip" makes a lot of sense: it's a UI pattern; you're styling the actual tooltips that show up and are definitely called tooltips by everyone.

Whereas, HTML works best when we try to keep things vague and general, so that the markup makes sense even when the page is being presented by a non-visual browser, or is reinterpreted for a new type of device (computer -> smartphone -> smartwatch), and so on. So a name like "hint" (or my old favorites according to the poll, "weak" and "transient") are aligned with HTML.

But in the end, I think I agree that (1) is more important than (2). If these things really behave the same, then probably they should get the same name. And probably "tooltip" is the better name. HTML has a long history of making compromises away from its pure platonic ideal, especially in naming. Examples include things like `<small>` and `<hr>`.

I do worry a bit that developers might be confused if a `<div popover="tooltip">` cannot be styled by `::tooltip { ... }`, and in fact is totally separate from it. But I trust you all to decide if that's realistic.

**Comment by keithamus** [comment]
I'm not sure how much weight my opinion carries but I am concerned we're conflating two quite separate things; `::tooltip` is for styling of the thing that `title=` produces, and `popover=hint` is an interaction pattern for exclusivity and ephemerality of the popover.

I _could_ make a `popover=hint` that behaves identically to `::tooltip`, as a replacement for it, and I might even name it the same thing. But the hint-style popover behaviour applies to many other things:

 - "rich tooltips" are usually given other names, such as popover, preview or hovercard. These are intentionally separated in design systems because they're more complex widgets and nest non-rich tooltips inside of them (but would allow nesting of themselves).
 - sub-menus would likely use `popover=hint` as it matches the interaction model, but they are definitely entirely separate from `::tooltip`.
 - error messages for things like form invalidation, often these might be called "flash" or "bubble". They are tooltip-like but 
 - toasts might also use the same pattern, and again might nest `::tooltip` but would definitely not nest toasts.

TL;DR inconsistency between these terms is a feature not a bug.

**Comment by hidde** [comment]
+1 to Keith, it makes sense to me to say “hint” is a broader category than tooltip. 

**Tooltip**, in [Tantek's 2000 suggestion](https://lists.w3.org/Archives/Public/www-style/2000Apr/0014.html), [Lea's 2009 proposal](https://lists.w3.org/Archives/Public/www-style/2009Apr/0225.html) and [the 2023 discussion](https://github.com/w3c/csswg-drafts/issues/8930), continues to mean the component that UAs display natively when `title` attribute is present, or components that look and behave pretty much like that. 

**Hint**, to me too, includes such tooltips, but also the interactions @keithamus mentions. I'd also add the “link preview” Wikipedia displays when you hover/focus a link.

I also think it's ok to have that one word for that broader category of items, given the alternative: being too specific. The popover attribute's name also covers a broader range of things, and (fwiw) in my conversations with developers about popover in the past year, that seemed to resonate with folks.

**Comment by bkardell** [comment]
+1. I agree with all of the points that @keithamus and @hidde made.

**Comment by annevk** [comment]
Given that the behavior is essentially a "tooltip" I still think that's the best name given the web platform is already using it (or will be using it). Having a lot of different names generally makes things harder to use. I don't think that the ability to use a tooltip for a variety of use cases detracts from that.

**Comment by keithamus** [comment]
I agree that lots of different names make things harder to use, especially where there is significant overlap. The opposite is also true though; naming incongruent behaviours the same makes things harder to use. In this case I believe the behaviour of `popover=hint` _is incongruent_; it overlaps but not entirely. The "tooltip" pattern is a superset of the behaviour prescribed by `popover=hint`.

I think naming this feature `popover=tooltip` will make it harder to use in cases where you don't want a tooltip (e.g. submenu), because semantics that `popover=tooltip` has will be assumed to be congruent with tooltips, but they aren't. Consequently if it were named `popover=tooltip`, I could foresee people asking questions like: "why does my `popover=tooltip` not open when I over over the `<button popovertarget>`" or "why does it not close when I mouseout of `<button popovertarget>`". These are expectations of `::tooltip` and the tooltip pattern in general, but they are not prescribed to `popover=hint`.

Ultimately I want the feature, it's important to me and I believe it'll improve web accessibility long term. If it has to be called `popover=tooltip` to get shipped, I'll live with that, it's a small price to pay. But considering this is a bikeshed issue I will maintain that, IMO, it is wrong name for this.

Much love to everyone, I know we all care about doing the right thing.

**Comment by mfreed7** [comment]
Since the `hint` naming is no longer the sticking point for `popover=hint`, I'm going to close this issue.
